### ISSUE TITLE
Remove need for GITHUB_TOKEN

### ISSUE THREAD

**Issue by StefanFabian** [cpp-linter/cpp-linter-action#55]
Hey there,
thanks a lot for this action!
It was unclear to me why this action wants the GITHUB_TOKEN since that can be quite a bit of a security risk.
A colleague of mine had a look and he thinks it isn't actually necessary.
Since he doesn't have time to make an issue I'll summarize his points here:

> That particular action could be written without the token without a problem. They use it to retrieve the changelist for the PR, which they could get easily from the local git repository just by having head and base ref.

For the annotations, you also don't need the token: see [the docs for setting an error message](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-error-message).

> You can just echo "::error file=app.js,line=1,col=5,endColumn=7::Missing semicolon" in your script and it will create an annotation in app.js, line 1 with "missing semicolon". Absolutely no need for API calls, and much easier this way.

Hope this helps you improve this action even further!

**Comment by 2bndy5** [comment]
TL;DR I'm open to suggested changes in the README, but not removing the token all together.

It is required to
1. get list of file changes for private repos
2. post thread comments. This action also removes duplicate previously posted thread comments (created by the bot). FWIW, we only remove a thread comment if the comment's body starts with a special HTML comment (`<!-- cpp-linter-action -->`) and the comment was created by the generic bot's user ID.

I agree that it could be a security risk (if you don't trust the src code openly available here). Note that the README does mention that the token is only required in the above situations. Furthermore, we include a link (in the README) to the GitHub docs about configuring the token (which also talks about or links to other docs about configuring the token's privileges).

Ultimately, this is a user decision. The token can be omitted if `changed-files-only` is false and `thread-comments` is false. `thread-comments` default value was recently changed to false after exploring private repo support (thread comments can't be supported for private repos).

**Comment by StefanFabian** [comment]
> 1. get list of file changes for private repos

Why don't you just compare the head with the base ref of the repository to get the file changes?
Is the git history not part of the checkout?

> if you don't trust the src code openly available here

I mean even if you do trust the code in its current state, most people will follow the readme and use the @master version of the action so you'd also have to trust that the repo is never compromised.
I don't think that's likely to happen but it seems like an unnecessary risk.

**Comment by 2bndy5** [comment]
> Is the git history not part of the checkout?

It is only if you set `fetch-depth` to `0` for `actions/checkout@v2+`. We're not trying to assume what the workflow does outside this action. Non-Ubuntu based runners are an exception due to how github allows us to use python to power this action.

We're also supporting push and PR events, so base refs aren't always going to be available.

It is true that the README should demonstrate using the latest release version tag, but that increases maintainability in the README. @shenxianpeng Are you ok with using `shenxianpeng/cpp-linter-action@v1` instead of ``shenxianpeng/cpp-linter-action@master` in the README?

**Comment by shenxianpeng** [comment]
@2bndy5 Of course no problem, I'll update it right away.

**Comment by 2bndy5** [comment]
Ok, the README now demonstrates using the `v1` tag. We also published a v1.4.2...

@StefanFabian There's also nothing stopping users from using a commit's SHA instead of a branch or tag. I recently helped a project that opted for using actions pinned to a specific commit (for the same security reason you mentioned).
